packages/samples/headless-react/src/components/category-facet/category-facet.class.tsx

```diff
@@ -1,34 +1,35 @@
-import {Component} from 'react';
+import {Component, ContextType} from 'react';
 import {
   buildCategoryFacet,
   CategoryFacet as HeadlessCategoryFacet,
   CategoryFacetState,
   CategoryFacetValue,
   Unsubscribe,
 } from '@coveo/headless';
-import {engine} from '../../engine';
 import {CategoryFacetSearch} from './category-facet-search';
+import {AppContext} from '../../context/engine';
 
 interface CategoryFacetProps {
   field: string;
   facetId: string;
 }
 
-export class CategoryFacet extends Component<CategoryFacetProps> {
-  private controller: HeadlessCategoryFacet;
-  public state: CategoryFacetState;
-  private unsubscribe: Unsubscribe = () => {};
+export class CategoryFacet extends Component<
+  CategoryFacetProps,
+  CategoryFacetState
+> {
+  static contextType = AppContext;
+  context!: ContextType<typeof AppContext>;
 
-  constructor(props: CategoryFacetProps) {
-    super(props);
+  private controller!: HeadlessCategoryFacet;
+  private unsubscribe: Unsubscribe = () => {};
 
-    this.controller = buildCategoryFacet(engine, {
-      options: {field: props.field, facetId: props.facetId},
+  componentDidMount() {
+    this.controller = buildCategoryFacet(this.context.engine!, {
+      options: {field: this.props.field, facetId: this.props.facetId},
     });
-    this.state = this.controller.state;
-  }
+    this.updateState();
```

Comment:
Due to the controller now being forced to be initialized in a lifecycle method, it's no longer possible to set the default state. A negative side-effect of this is that `render` is going to be called twice: once with `this.state` and `this.controller` being `undefined`, then once with both of those being set to their proper value.